Tasks/Maven/maven2.js

```diff
@@ -48,40 +51,149 @@ var publishJUnitResults = tl.getInput('publishJUnitResults');
 var testResultsFiles = tl.getInput('testResultsFiles', true);
 
 function publishTestResults(publishJUnitResults, testResultsFiles) {
-  if(publishJUnitResults == 'true') {
-    //check for pattern in testResultsFiles
-    if(testResultsFiles.indexOf('*') >= 0 || testResultsFiles.indexOf('?') >= 0) {
-      tl.debug('Pattern found in testResultsFiles parameter');
-      var buildFolder = tl.getVariable('agent.buildDirectory');
-      var allFiles = tl.find(buildFolder);
-      var matchingTestResultsFiles = tl.match(allFiles, testResultsFiles, { matchBase: true });
+    if (publishJUnitResults == 'true') {
+        //check for pattern in testResultsFiles
+        if (testResultsFiles.indexOf('*') >= 0 || testResultsFiles.indexOf('?') >= 0) {
+            tl.debug('Pattern found in testResultsFiles parameter');
+            var buildFolder = tl.getVariable('agent.buildDirectory');
+            var allFiles = tl.find(buildFolder);
+            var matchingTestResultsFiles = tl.match(allFiles, testResultsFiles, {
+                matchBase: true
+            });
+        } else {
+            tl.debug('No pattern found in testResultsFiles parameter');
+            var matchingTestResultsFiles = [testResultsFiles];
+        }
+
+        if (!matchingTestResultsFiles) {
+            tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
+            return 0;
+        }
+
+        var tp = new tl.TestPublisher("JUnit");
+        tp.publish(matchingTestResultsFiles, false, "", "");
     }
-    else {
-      tl.debug('No pattern found in testResultsFiles parameter');
-      var matchingTestResultsFiles = [testResultsFiles];
+}
+
```

Comment:
The sonar:sonar goal cannot run with the other goals due to a SonarQube limitation / bug - so a separate maven execution is needed